python-checks/src/main/java/org/sonar/python/checks/ConfusingTypeCheckingCheck.java

```diff
@@ -106,4 +109,38 @@ public String message(@Nullable String name, String missingMethod) {
       return String.format("Fix this \"%s\" operation; Previous type checks suggest that this expression does not have this method.", missingMethod);
     }
   }
+
+  private static class IterationOnNonIterableCheck extends IterationOnNonIterable {
+
+    @Override
+    boolean isValidIterable(Expression expression, List<LocationInFile> secondaries) {
+      InferredType type = expression.type();
+      secondaries.add(InferredTypes.typeClassLocation(type));
+      return !containsDeclaredType(type) || type.declaresMember("__iter__") || type.declaresMember("__getitem__");
+    }
+
+    @Override
+    String message(Expression expression, boolean isForLoop) {
+      String typeName = InferredTypes.typeName(expression.type());
+      String expressionName = nameFromExpression(expression);
+      String expressionNameString = expressionName != null ? String.format("\"%s\"", expressionName) : "it";
+      String typeNameString = typeName != null ? String.format("has type \"%s\" and", typeName) : "";
+      return isForLoop && isAsyncIterable(expression) ?
+        String.format("Add \"async\" before \"for\"; Previous type checks suggest that %s %s is an async generator.", expressionNameString, typeNameString) :
+        String.format("Replace this expression; Previous type checks suggest that %s %s isn't iterable.", expressionNameString, typeNameString);
+    }
+
+    @Override
+    boolean isAsyncIterable(Expression expression) {
+      InferredType type = expression.type();
+      return type.declaresMember("__aiter__");
```

Comment:
Here we don't need to check `containsDeclaredType` because it's called only in method `message` right? Can you add a comment for that?